shared/domains/chelonia/chelonia.js

```diff
@@ -108,6 +226,15 @@ export default (sbp('sbp/selectors/register', {
       return stack
     }
   },
+  'chelonia/withEnv': async function (contractID: string, env: Object, sbpInvocation: Array<*>) {
+    const savedEnv = this.env
+    this.env = env
+    try {
+      return await sbp('okTurtles.eventQueue/queueEvent', `chelonia/withEnv/${contractID}`, sbpInvocation)
```

Comment:
@corrideat I'm worried that using \``chelonia/withEnv/${contractID}`\` as the key may be dangerous and lead to unpredictable behavior.

We want to make sure that any actions being performed on a contract happen serially right? If we don't use `contractID`, then a withEnv invocation could happen at the same as when we're processing a new message via handleEvent, or during the processing of a contractSync, and in either case this can result in messages being processed out-of-order right? And if that happens then the entire system breaks due to exceptions being thrown.

Or am I missing something?

Reply:
Since this is rather exceptional and non-intuitive, could you please add a detailed comment here explaining this?

Reply:
Thinking about this, the correct solution is probably not this. As mentioned, `withEnv` needs to be protected and it's a different queue from the other events. However, the issue is that `.env` is implemented in shared state, which can result in race conditions like you mention.
I also don't think something like `withEnv` can be _completely_ avoided, but I suggest a new design:

1. Get rid of `.env` / `withEnv`
2. Pass environment as an argument to invocations. All invocations to chelonia selectors take this argument and pass it along (possibly modified) to further calls they make.

This is more explicit and robust

Reply:
@corrideat Let's do a call about this

Reply:
To summarize call, the solution we've come up with for now is to put the current contents of `'chelonia/withEnv'` inside of a serialized queue using `queueEvent`, using the same string as we're currently using for the queue name (`chelonia/withEnv/${contractID}`), and then inside of that simply directly call the `sbpInvocation` (instead of creating another queue)

Reply:
This was (hopefully) resolved in: https://github.com/okTurtles/group-income/pull/1536